### DISCUSSION TITLE
How to build a Map-Reduce Summarization Chain / Workflow?

### DISCUSSION THREAD

**Discussion by djpecot** [FlowiseAI/Flowise#835] [Q&A]
Howdy! Looking to start using this tool in addition to Langchain for quick prototyping :)

What would be the best practice to implement a [map-reduce style summarization of an entire document](https://python.langchain.com/docs/use_cases/summarization#option-2-map-reduce) (like in Langchain)? Is there a best practice or way to perform any kind of custom summarization task on a document that exceeds the token limit for a single prompt? I browsed the docs, checked the UI, and checked issues and discussions here without finding anything.

**Comment by HenryHengZJ** [comment]
In ConveraationalRetrievalQAChain, click the Additional Parameters, from there you can select stuff, map reduce type

**Reply by djpecot** [reply]
Yep, I got this one working, however seems like it's just pulling random rows from my vectorstore and not giving an appropriate summary of the entire pdf I uploaded (just one): 

I understand that the k value also can change how many documents are returned, however in this case I want to perform a map-reduce summarization chain across my entire document (I mean pdf in this case, not how a document is chunked).
<img width="1534" alt="Screenshot 2023-08-26 at 8 59 09 PM" src="https://github.com/FlowiseAI/Flowise/assets/8185181/e34162ac-6c82-422d-9577-48a53985e78e">
![image](https://github.com/FlowiseAI/Flowise/assets/8185181/957b9f91-166c-48c6-ac31-8a277562501a)

**Reply by djpecot** [reply]
Here's another example: 
![image](https://github.com/FlowiseAI/Flowise/assets/8185181/f704a759-1fbb-4c14-b54d-3d17938690f7)

**Comment by djpecot** [comment]
Random Thought:

I guess this process of iterating over each part of a document would be similar to the creation [hypothetical document embeddings](https://python.langchain.com/docs/use_cases/question_answering/how_to/hyde#:~:text=At%20a%20high%20level%2C%20HyDE,used%20to%20generate%20those%20documents.) as well in that the LLM must interact with all available data, but not within the same completion.

**Comment by HenryHengZJ** [comment]
oh got it, you are trying to use summarization chain with map-reduce, instead of a conversational retrieval qa chain?

for now we still haven't integrated into flowise yet, but might be a good addition - https://js.langchain.com/docs/modules/chains/popular/summarize

**Reply by djpecot** [reply]
@HenryHengZJ bingo. I haven't dug into Flowwise too much yet, but is there a code runtime I can use to write custom langchain code? If I could do that, wouldn't necessarily need as a feature since I can write it myself :)

**Reply by kianwoon** [reply]
I'm also interested in a document summarisation chain for LangChain. It would be a valuable feature to add. While LangFlow doesn't currently offer this capability, they do support custom components through their beta feature, which is worth considering.

**Comment by luizcarvalho** [comment]
@djpecot Do you solved this question?
